// flowController/ss_io.js
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

IO = {
    createResponse : function(){
        var parent = {
            __version : 1,
            __code : null,
            __messages : [],
            __errors : [],
            __status : "complete",
            __instanceId : null
        };
        var child = {
            getVersion : function(){
                return this.__version;
            },
            getCode : function(){
                return this.__code;
            },
            setCode : function(code){
                this.__code = code;
                return this;
            },
            getMessages : function(){
                return this.__messages;
            },
            addMessage : function(message){
                if(!Array.isArray(this.__messages)){
                    this.prototype.__messages = new Array();
                }
                this.__messages.push(message);
                return this;
            },
            addError : function(error){
                if(!Array.isArray(this.__errors)){
                    this.prototype.__errors = new Array();
                }
                this.__code = 500;
                this.__errors.push(error);
                return this;
            },
            getErrors : function(){
                return this.__errors;
            },
            getStatus : function(){
                return this.__status;
            },
            setStatus : function(status){
                if(typeof status === "string"){
                    this.__status = status;
                }else{
                    this.__status = null;
                }
                return this;
            },
            getInstanceId : function(){
                return this.__instanceId;
            },
            setInstanceId : function(id){
                this.__instanceId = id;
                return this;
            },
            toObject : function(){
                return {
                    version : child.getVersion(),
                    code : child.getCode(),
                    messages : child.getMessages(),
                    errors : child.getErrors(),
                    status : child.getStatus(),
                    instanceId : child.getInstanceId()
                };
            }
        };
        child.__proto__ = parent;
        return child;
    }
};

// Functions which will be available to external callers
module.exports = IO;
